### PR TITLE
2 create members database

### LINKED ISSUE #2: Create members database
In the `members` app add the following tables to `models.py`. Django documentation [here](https://docs.djangoproject.com/en/3.2/topics/db/models/).

image

Field constraints (docs [here](https://docs.djangoproject.com/en/3.2/ref/models/fields/)):
- `UUID` is primary key 
- `username` and `email` should be marked as unique
- `username`, `password`, `active` and `email` should be marked as not null
- `username` is foreign key in both tables
- also note the max length in case of varchar attributes
- use on_delete: cascade

NOTE: The passwords will be hashed, NOT stored in plain text!
Edit: Add a [UUIDField](https://www.geeksforgeeks.org/uuidfield-django-models/) to the users table

### LINKED ISSUE #3: Set up an admin profile for the members app
In the `members` app create an admin profile in `admin.py`. Docs [here](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/). The admin should be able to add new members to the `members table` defined in `models.py` (#2) and change the fields (role). To test if this works you might need to do some [migrations](https://docs.djangoproject.com/en/3.2/topics/migrations/).

### LINKED ISSUE #4: Add a register page to the members app
# Frontend 
- add a HTML form with 3 fields: username, password, confirm password.
- check if new password is the same as confirm new password.
- use this regex to check if the password contains minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character:

`"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$"` 

Yes I know, it's very ugly. Source [here](https://stackoverflow.com/questions/19605150/regex-for-password-must-contain-at-least-eight-characters-at-least-one-number-a). If the password is okay, send the POST request otherwise display an error message.

# Backend 
- handle a POST request with the username, password.
- if there is an entry with the input username in the `members table`, respond with an error message: username already in use or something along those lines...
- otherwise add a new entry for the user to the `members table` and `passwords table`. 

We'll use role `radio_amateur` as default for new users and role `operator` for our GS crew (manually changed by admin).

### LINKED ISSUE #5: Add a change password page to the members app
# Frontend 
- add a HTML form with 3 fields: old password, new password, confirm new password.
- check if new password is the same as the confirmed new password.
- use this regex to check if the new password contains minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character:

`"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$"` 

Source [here](https://stackoverflow.com/questions/19605150/regex-for-password-must-contain-at-least-eight-characters-at-least-one-number-a). If the password is okay, send the POST request otherwise display an error message.

NOTE:  the user should already be logged in and authenticated when the password is reset

# Backend 
- handle the POST request and update the password.

### LINKED ISSUE #7: Members app (user management)
Check out [this](https://towardsdatascience.com/build-a-user-authentication-web-app-with-python-and-django-c60000148724) article, looks useful!
- [x] #2 
- [x] #3 
- [x] #4 
- [x] #6 
- [x] #5 
- [x] #44
- [x] #32
- [x] #33  
- [x] #31 
- [x] #35 
- [x] #28 
- [x] #29 
- [x] #34

### PR DESCRIPTION
**Pull request by chaiwon59**
Close #2 
Close #3 
Close #4 
Close #5

### CODE REVIEW

**Comment by iiacoban42** [comment]
Good job on getting the login to work! 👍🏻 Password change also works nicely! Couldn't load `http://127.0.0.1:8000/members/set/` though. Also can you change `set_password` from `views.py` to register a new user into the database, I think that was supposed to handle #4. And please use <input type="password"> for the password field to hide the letters. I'll come back with more comments later this week, need a bit of time to test it out.

**Comment by StefanoSperetta** [comment]
Would it be possible to add a password / username recovery page? Maybe simply typing the email address in as input, triggering an email with a recovery link such that a new password can be set, and providing also the username?

Personally, I would not do it as part of this issue but I would add it as a separate one so we can close this one.

File: src/members/migrations/0001_initial.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                ('is_staff', models.BooleanField(default=False, help_text='Designates whether the user can log into this admin site.', verbose_name='staff status')),
+                ('is_active', models.BooleanField(default=True, help_text='Designates whether this user should be treated as active. Unselect this instead of deleting accounts.', verbose_name='active')),
+                ('date_joined', models.DateTimeField(default=django.utils.timezone.now, verbose_name='date joined')),
+                ('UUID', models.UUIDField(default=uuid.uuid4, editable=False, primary_key=True, serialize=False)),
+                ('username', models.CharField(max_length=70, unique=True)),
+                ('email', models.CharField(max_length=320, unique=True)),
+                ('password', models.CharField(blank=True, max_length=120, null=True, validators=[django.core.validators.RegexValidator(message='Enter password containing min 8 characters, min 1 upper and lower case letter, 1 number and 1 special character', regex='^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$')])),
```
Comment (NotUnlikeTheWaves) [inline src/members/migrations/0001_initial.py:31]: Does this mean that the regex validation applies to the database field?

**Comment by NotUnlikeTheWaves** [comment]
Out of scope for this pull request but if we're using passwords we should try to salt them as well

**Comment by NotUnlikeTheWaves** [comment]
Isn't implementing a members object done in #16 too?

File: src/members/templates/members/set/set_password.html [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+<!--    <form action="/members/set/" method="post">-->
+<!--        <label for="password">Password </label>-->
+<!--        <input id="password" type="text" name="password" value="{{ current_password }}">-->
+<!--        <input type="submit" value="OK">-->
+<!--    </form>-->
+<!--    <form action="/members/set/" method="post">-->
+<!--        <label for="confirm_password">Password </label>-->
+<!--        <input id="confirm_password" type="text" name="confirm_password" value="{{ current_confirm_password }}">-->
+<!--        <input type="submit" value="OK">-->
+<!--    </form>-->
```
Comment (iiacoban42) [inline src/members/templates/members/set/set_password.html:36]: If you think this will be useful again keep it, if not please remove it.

**Review by iiacoban42 — COMMENTED** [review]
For the first iteration of the login system, this is quite good! The basic functionality is here and it works! Kudos to you Chaiwon! Have look at the comments and try to fix pylint, it ignores tests and migrations now. In the meantime I'll add some new issues to build on top of this first iteration. I think it's best if merge this one and then extend the functionality down the line.

File: src/members/models.py [hunk 3]
```diff
@@ -1,3 +1,30 @@
...
+
+    UUID = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    username = models.CharField(max_length=70, unique=True)
+    email = models.CharField(max_length=320, unique=True, null=False, blank=False)
+    password = models.CharField(max_length=120,
+                                validators=[
+                                    RegexValidator(
+                                        regex='^(?=.*[a-z])(?=.*[A-Z])'
+                                              '(?=.*\\d)(?=.*[@$!%*?&])'
+                                              '[A-Za-z\\d@$!%*?&]{8,}$',
```
Comment (NotUnlikeTheWaves) [inline src/members/models.py:18]: I'm not sure this validator works for the input through the website, since I can use 'testpass' as a password.

**Comment by NotUnlikeTheWaves** [comment]
I've merged this with the main branch and it works nicely with the postgres db, very cool to see. I've pushed it with the fixed merge conflict. Maybe we can merge this PR and take a look at the remaining points in this in a new PR, so other issues can benefit from working with the members objects?

**Comment by NotUnlikeTheWaves** [comment]
Well that took a while to get through the linter :D

**Comment by NotUnlikeTheWaves** [comment]
I'm still not sure we need the custom authentication backend, I feel we should be able to let django talk to the DB ourselves

**Comment by iiacoban42** [comment]
Good point @NotUnlikeTheWaves! Thanks! I'll merge this one!

**Review by iiacoban42 — APPROVED** [review]

**Pull request merged**